File: core/src/main/java/decoder/NativeDecoder.java
package decoder;

import java.lang.ref.Cleaner;

public class NativeDecoder implements AutoCloseable {
    private final long nativeHandle;
    private final Cleaner.Cleanable cleanable;

    private static final Cleaner cleaner = Cleaner.create();

    private record CleanupAction(long handle) implements Runnable {
        @Override
        public void run() {
            deleteNative(handle);
        }
    }

    public NativeDecoder(String location, boolean findAudioStream, boolean findVideoStream) throws Exception {
        long nativeHandle = createNative(location, findAudioStream, findVideoStream);
        if (nativeHandle == 0) {
            throw new Exception("Unable to instantiate NativeDecoder");
        }
        this.nativeHandle = nativeHandle;
        this.cleanable = cleaner.register(this, new CleanupAction(nativeHandle));
    }

    private static native long createNative(String location, boolean findAudioStream, boolean findVideoStream);

    private static native NativeFormat getFormatNative(long handle);

    private static native NativeFrame nextFrameNative(long handle, int width, int height);

    private static native void seekToNative(long handle, long timestampMicros, boolean keyframesOnly);

    private static native void resetNative(long handle);

    private static native void deleteNative(long handle);

    public NativeFormat getFormat() {
        return getFormatNative(nativeHandle);
    }

    public NativeFrame nextFrame(Integer width, Integer height) {
        return nextFrameNative(nativeHandle, width == null ? getFormat().width() : width, height == null ? getFormat().height() : height);
    }

    public void seekTo(long timestampMicros, boolean keyframesOnly) {
        seekToNative(nativeHandle, timestampMicros, keyframesOnly);
    }

    public void reset() {
        resetNative(nativeHandle);
    }

    @Override
    public void close() {
        cleanable.clean();
    }
}

File: core/src/main/kotlin/com/github/numq/klarity/core/decoder/NativeDecoder.kt
package com.github.numq.klarity.core.decoder

import com.github.numq.klarity.core.cleaner.NativeCleaner
import com.github.numq.klarity.core.format.NativeFormat
import com.github.numq.klarity.core.frame.NativeFrame
import java.io.Closeable
import java.util.concurrent.atomic.AtomicLong

internal class NativeDecoder(
    location: String,
    findAudioStream: Boolean,
    findVideoStream: Boolean,
    decodeAudioStream: Boolean,
    decodeVideoStream: Boolean,
    hardwareAccelerationCandidates: IntArray? = null,
) : Closeable {
    private object Native {
        @JvmStatic
        external fun getAvailableHardwareAcceleration(): IntArray?

        @JvmStatic
        external fun create(
            location: String,
            findAudioStream: Boolean,
            findVideoStream: Boolean,
            decodeAudioStream: Boolean,
            decodeVideoStream: Boolean,
            hardwareAccelerationCandidates: IntArray
        ): Long

        @JvmStatic
        external fun getFormat(handle: Long): NativeFormat

        @JvmStatic
        external fun decodeAudio(handle: Long, buffer: Long, capacity: Int): NativeFrame?

        @JvmStatic
        external fun decodeVideo(handle: Long, buffer: Long, capacity: Int): NativeFrame?

        @JvmStatic
        external fun seekTo(handle: Long, timestampMicros: Long, keyframesOnly: Boolean): Long

        @JvmStatic
        external fun reset(handle: Long)

        @JvmStatic
        external fun delete(handle: Long)
    }

    companion object {
        fun getAvailableHardwareAcceleration() = Native.getAvailableHardwareAcceleration() ?: intArrayOf()
    }

    private val lock = Any()

    private val nativeHandle = AtomicLong(-1L)

    private fun ensureOpen() {
        check(nativeHandle.get() != -1L) { "Native sampler is closed" }
    }

    init {
        synchronized(lock) {
            nativeHandle.set(
                Native.create(
                    location = location,
                    findAudioStream = findAudioStream,
                    findVideoStream = findVideoStream,
                    decodeAudioStream = decodeAudioStream,
                    decodeVideoStream = decodeVideoStream,
                    hardwareAccelerationCandidates = hardwareAccelerationCandidates ?: intArrayOf()
                )
            )

            require(nativeHandle.get() != -1L) { "Could not instantiate native decoder" }
        }
    }

    private val cleanable = NativeCleaner.cleaner.register(this) {
        synchronized(lock) {
            ensureOpen()

            Native.delete(handle = nativeHandle.get())

            nativeHandle.set(-1L)
        }
    }

    val format = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.getFormat(handle = nativeHandle.get())
        }
    }

    fun getNativeHandle() = nativeHandle.get()

    fun decodeAudio(buffer: Long, capacity: Int) = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.decodeAudio(handle = nativeHandle.get(), buffer = buffer, capacity = capacity)
        }
    }

    fun decodeVideo(buffer: Long, capacity: Int) = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.decodeVideo(handle = nativeHandle.get(), buffer = buffer, capacity = capacity)
        }
    }

    fun seekTo(timestampMicros: Long, keyframesOnly: Boolean) = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.seekTo(
                handle = nativeHandle.get(), timestampMicros = timestampMicros, keyframesOnly = keyframesOnly
            ).takeIf { it >= 0 } ?: timestampMicros
        }
    }

    fun reset() = synchronized(lock) {
        runCatching {
            ensureOpen()

            Native.reset(handle = nativeHandle.get())
        }
    }

    override fun close() = cleanable.clean()
}